lightning/src/chain/chainmonitor.rs

```diff
@@ -162,39 +184,32 @@ impl<ChanSigner: ChannelKeys, C: Deref, T: Deref, F: Deref, L: Deref> ChainMonit
 		Ok(())
 	}
 
-	/// Updates the monitor that watches the channel referred to by the given outpoint.
-	fn update_monitor(&self, outpoint: OutPoint, update: ChannelMonitorUpdate) -> Result<(), MonitorUpdateError> {
+	fn update_channel(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr> {
+		// Update the monitor that watches the channel referred to by the given outpoint.
 		let mut monitors = self.monitors.lock().unwrap();
-		match monitors.get_mut(&outpoint) {
+		match monitors.get_mut(&funding_txo) {
+			None => {
+				log_error!(self.logger, "Failed to update channel monitor: no such monitor registered");
+				return Err(ChannelMonitorUpdateErr::PermanentFailure);
+			},
 			Some(orig_monitor) => {
 				log_trace!(self.logger, "Updating Channel Monitor for channel {}", log_funding_info!(orig_monitor));
-				orig_monitor.update_monitor(update, &self.broadcaster, &self.logger)
-			},
-			None => Err(MonitorUpdateError("No such monitor registered"))
-		}
-	}
-}
-
-impl<ChanSigner: ChannelKeys, C: Deref + Sync + Send, T: Deref + Sync + Send, F: Deref + Sync + Send, L: Deref + Sync + Send> chain::Watch for ChainMonitor<ChanSigner, C, T, F, L>
-	where C::Target: chain::Filter,
-	      T::Target: BroadcasterInterface,
-	      F::Target: FeeEstimator,
-	      L::Target: Logger,
-{
-	type Keys = ChanSigner;
-
-	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<ChanSigner>) -> Result<(), ChannelMonitorUpdateErr> {
-		match self.add_monitor(funding_txo, monitor) {
-			Ok(_) => Ok(()),
-			Err(_) => Err(ChannelMonitorUpdateErr::PermanentFailure),
-		}
-	}
-
-	fn update_channel(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr> {
-		match self.update_monitor(funding_txo, update) {
-			Ok(_) => Ok(()),
-			Err(_) => Err(ChannelMonitorUpdateErr::PermanentFailure),
+				let res = orig_monitor.update_monitor(&update, &self.broadcaster, &self.logger);
+				// Even if updating the monitor returns an error, the monitor's state will
+				// still be changed. So, persist the updated monitor despite the error.
+				if let Err(e) = self.persister.update_persisted_channel(funding_txo, &update, orig_monitor) {
+					if !res.is_err() {
+						log_error!(self.logger, "Failed to persist new channel data: {:?}", e);
```

Comment:
Do we want to log this regardless of whether the update failed? Maybe it would be best to perform both operations before logging. Then log each error result independently before finally returning if either is an error.

Reply:
> Maybe it would be best to perform both operations before logging. Then log each error result independently before finally returning if either is an error.

iiuc I did this, and I think it's better. Lmk if it still needs work!